tests/shell/tests.py

```diff
@@ -94,3 +122,77 @@ def test_shell_with_bpython_not_installed(self, select):
     # returns EOF and so select always shows that sys.stdin is ready to read.
     # This causes problems because of the call to select.select() toward the
     # end of shell's handle() method.
+
+    def test_get_namespace(self):
+        with self.settings(
+            INSTALLED_APPS=[
+                "shell",
+                "django.contrib.auth",
+                "django.contrib.contenttypes",
+            ]
+        ):
+
+            cmd = shell.Command()
+            namespace = cmd.get_namespace()
+
+            self.assertEqual(
+                namespace,
+                {
+                    "Marker": shell_models.Marker,
+                    "Phone": shell_models.Phone,
+                    "ContentType": ContentType,
+                    "Group": Group,
+                    "Permission": Permission,
+                    "User": User,
+                    "auth_models": auth_model_module,
+                    "contenttypes_models": contenttypes_model_module,
+                    "shell_models": shell_models,
+                },
+            )
+
+    def test_get_namespace_precedence(self):
+        class Marker(models.Model):
+            class Meta:
+                app_label = "basic"
+
+        with self.settings(INSTALLED_APPS=["basic", "shell"]):
+            cmd = shell.Command()
+            namespace = cmd.get_namespace()
+
+            self.assertTrue(Marker in namespace.values())
```

Comment:
Use specific assertion functions so the message is better on failure (for all below assertions)
```suggestion
            self.assertIn(Marker, namespace.values())
```